File: packages/usage-spec/src/kdl.ts
import { Document, Node, format, parse } from "@bgotink/kdl";
import type { Spec, SpecArg, SpecFlag, SpecCommand, SpecChoices } from "./spec.js";

function buildChoicesNode(choices: SpecChoices): Node {
  const node = Node.create("choices");
  for (const c of choices.values) {
    node.addArgument(c);
  }
  return node;
}

function buildArgNode(arg: SpecArg): Node {
  // Build usage string: <required> or [optional], with … for variadic
  let usage: string;
  if (arg.required) {
    usage = `<${arg.name}>`;
  } else {
    usage = `[${arg.name}]`;
  }
  if (arg.var) {
    usage += "\u2026";
  }

  const node = Node.create("arg");
  node.addArgument(usage);

  if (arg.help) node.setProperty("help", arg.help);
  if (!arg.required) node.setProperty("required", false);
  if (arg.var) node.setProperty("var", true);
  if (arg.hide) node.setProperty("hide", true);
  if (arg.default.length === 1) node.setProperty("default", arg.default[0]);

  const hasChildren = arg.choices != null || arg.default.length > 1;
  if (hasChildren) {
    const children = new Document();

    if (arg.default.length > 1) {
      const defaultNode = Node.create("default");
      for (const d of arg.default) {
        defaultNode.addArgument(d);
      }
      children.appendNode(defaultNode);
    }

    if (arg.choices) {
      children.appendNode(buildChoicesNode(arg.choices));
    }

    node.children = children;
  }

  return node;
}

function buildFlagArgNode(arg: SpecArg): Node {
  const node = Node.create("arg");
  node.addArgument(`<${arg.name}>`);

  if (arg.help) node.setProperty("help", arg.help);

  if (arg.choices) {
    const children = new Document();
    children.appendNode(buildChoicesNode(arg.choices));
    node.children = children;
  }

  return node;
}

function buildFlagNode(flag: SpecFlag): Node {
  // Build the flag name: "-s --long"
  const nameParts: string[] = [];
  if (flag.short) nameParts.push(`-${flag.short}`);
  if (flag.long) nameParts.push(`--${flag.long}`);
  const flagName = nameParts.join(" ");

  const node = Node.create("flag");
  node.addArgument(flagName);

  if (flag.help) node.setProperty("help", flag.help);
  if (flag.required) node.setProperty("required", true);
  if (flag.var) node.setProperty("var", true);
  if (flag.hide) node.setProperty("hide", true);
  if (flag.global) node.setProperty("global", true);
  if (flag.count) node.setProperty("count", true);
  if (flag.negate) node.setProperty("negate", flag.negate);
  if (flag.deprecated) node.setProperty("deprecated", flag.deprecated);
  if (flag.default.length === 1) node.setProperty("default", flag.default[0]);
  else if (flag.defaultBool != null) node.setProperty("default", flag.defaultBool);
  if (flag.env) node.setProperty("env", flag.env);

  const hasChildren =
    flag.helpLong !== "" || flag.arg != null || flag.default.length > 1;

  if (hasChildren) {
    const children = new Document();

    if (flag.helpLong) {
      const longHelpNode = Node.create("long_help");
      longHelpNode.addArgument(flag.helpLong);
      children.appendNode(longHelpNode);
    }

    if (flag.default.length > 1) {
      const defaultNode = Node.create("default");
      for (const d of flag.default) {
        defaultNode.addArgument(d);
      }
      children.appendNode(defaultNode);
    }

    if (flag.arg) {
      children.appendNode(buildFlagArgNode(flag.arg));
    }

    node.children = children;
  }

  return node;
}

function buildCommandNode(cmd: SpecCommand): Node {
  const node = Node.create("cmd");
  node.addArgument(cmd.name);

  if (cmd.hide) node.setProperty("hide", true);
  if (cmd.subcommandRequired) node.setProperty("subcommand_required", true);
  if (cmd.help) node.setProperty("help", cmd.help);
  if (cmd.deprecated) node.setProperty("deprecated", cmd.deprecated);

  const hasChildren =
    cmd.helpLong !== "" ||
    cmd.aliases.length > 0 ||
    cmd.flags.length > 0 ||
    cmd.args.length > 0 ||
    cmd.cmds.length > 0;

  if (hasChildren) {
    const children = new Document();

    if (cmd.aliases.length > 0) {
      const aliasNode = Node.create("alias");
      for (const a of cmd.aliases) {
        aliasNode.addArgument(a);
      }
      children.appendNode(aliasNode);
    }

    if (cmd.helpLong) {
      const longHelpNode = Node.create("long_help");
      longHelpNode.addArgument(cmd.helpLong);
      children.appendNode(longHelpNode);
    }

    for (const flag of cmd.flags) {
      children.appendNode(buildFlagNode(flag));
    }

    for (const arg of cmd.args) {
      children.appendNode(buildArgNode(arg));
    }

    for (const sub of cmd.cmds) {
      children.appendNode(buildCommandNode(sub));
    }

    node.children = children;
  }

  return node;
}

export function renderKDL(spec: Spec): string {
  const doc = new Document();

  if (spec.name) {
    const nameNode = Node.create("name");
    nameNode.addArgument(spec.name);
    doc.appendNode(nameNode);
  }

  if (spec.bin) {
    const binNode = Node.create("bin");
    binNode.addArgument(spec.bin);
    doc.appendNode(binNode);
  }

  if (spec.version) {
    const versionNode = Node.create("version");
    versionNode.addArgument(spec.version);
    doc.appendNode(versionNode);
  }

  if (spec.about) {
    const aboutNode = Node.create("about");
    aboutNode.addArgument(spec.about);
    doc.appendNode(aboutNode);
  }

  if (spec.long) {
    const longAboutNode = Node.create("long_about");
    longAboutNode.addArgument(spec.long);
    doc.appendNode(longAboutNode);
  }

  if (spec.usage) {
    const usageNode = Node.create("usage");
    usageNode.addArgument(spec.usage);
    doc.appendNode(usageNode);
  }

  for (const flag of spec.flags) {
    doc.appendNode(buildFlagNode(flag));
  }

  for (const arg of spec.args) {
    doc.appendNode(buildArgNode(arg));
  }

  for (const cmd of spec.cmds) {
    doc.appendNode(buildCommandNode(cmd));
  }

  return format(doc);
}

/**
 * Validates KDL output by parsing it back.
 * Throws if the KDL is invalid.
 */
export function validateKDL(kdl: string): void {
  parse(kdl);
}

File: packages/usage-spec-python/src/usage_spec/kdl.py
"""KDL renderer for usage spec, aligned with @usage-spec/core kdl.ts output."""

from __future__ import annotations

import ckdl

from .spec import Spec, SpecArg, SpecFlag, SpecCommand, SpecChoices


def _node(name: str, args: list[str] | None = None, properties: dict | None = None, children: list[ckdl.Node] | None = None) -> ckdl.Node:
    n = ckdl.Node(name)
    if args:
        n.args = args
    if properties:
        n.properties = properties
    if children:
        n.children = children
    return n


def _build_choices_node(choices: SpecChoices) -> ckdl.Node:
    return _node("choices", args=choices.values)


def _build_arg_node(arg: SpecArg) -> ckdl.Node:
    usage = f"<{arg.name}>" if arg.required else f"[{arg.name}]"
    if arg.var:
        usage += "\u2026"

    properties: dict = {}
    if arg.help:
        properties["help"] = arg.help
    if not arg.required:
        properties["required"] = False
    if arg.var:
        properties["var"] = True
    if arg.hide:
        properties["hide"] = True
    if len(arg.default) == 1:
        properties["default"] = arg.default[0]

    children: list[ckdl.Node] = []
    if len(arg.default) > 1:
        children.append(_node("default", args=arg.default))
    if arg.choices:
        children.append(_build_choices_node(arg.choices))

    return _node("arg", args=[usage], properties=properties or None, children=children or None)


def _build_flag_arg_node(arg: SpecArg) -> ckdl.Node:
    properties: dict = {}
    if arg.help:
        properties["help"] = arg.help

    children: list[ckdl.Node] = []
    if arg.choices:
        children.append(_build_choices_node(arg.choices))

    return _node("arg", args=[f"<{arg.name}>"], properties=properties or None, children=children or None)


def _build_flag_node(flag: SpecFlag) -> ckdl.Node:
    name_parts: list[str] = []
    if flag.short:
        name_parts.append(f"-{flag.short}")
    if flag.long:
        name_parts.append(f"--{flag.long}")
    flag_name = " ".join(name_parts)

    properties: dict = {}
    if flag.help:
        properties["help"] = flag.help
    if flag.required:
        properties["required"] = True
    if flag.var:
        properties["var"] = True
    if flag.hide:
        properties["hide"] = True
    if flag.global_:
        properties["global"] = True
    if flag.count:
        properties["count"] = True
    if flag.negate:
        properties["negate"] = flag.negate
    if flag.deprecated:
        properties["deprecated"] = flag.deprecated
    if len(flag.default) == 1:
        properties["default"] = flag.default[0]
    elif flag.default_bool is not None:
        properties["default"] = flag.default_bool
    if flag.env:
        properties["env"] = flag.env

    children: list[ckdl.Node] = []
    if flag.help_long:
        children.append(_node("long_help", args=[flag.help_long]))
    if len(flag.default) > 1:
        children.append(_node("default", args=flag.default))
    if flag.arg:
        children.append(_build_flag_arg_node(flag.arg))

    return _node("flag", args=[flag_name], properties=properties or None, children=children or None)


def _build_command_node(cmd: SpecCommand) -> ckdl.Node:
    properties: dict = {}
    if cmd.hide:
        properties["hide"] = True
    if cmd.subcommand_required:
        properties["subcommand_required"] = True
    if cmd.help:
        properties["help"] = cmd.help
    if cmd.deprecated:
        properties["deprecated"] = cmd.deprecated

    children: list[ckdl.Node] = []
    if cmd.aliases:
        children.append(_node("alias", args=cmd.aliases))
    if cmd.help_long:
        children.append(_node("long_help", args=[cmd.help_long]))
    for flag in cmd.flags:
        children.append(_build_flag_node(flag))
    for arg in cmd.args:
        children.append(_build_arg_node(arg))
    for sub in cmd.cmds:
        children.append(_build_command_node(sub))

    return _node("cmd", args=[cmd.name], properties=properties or None, children=children or None)


def render_kdl(spec: Spec) -> str:
    nodes: list[ckdl.Node] = []

    if spec.name:
        nodes.append(_node("name", args=[spec.name]))
    if spec.bin:
        nodes.append(_node("bin", args=[spec.bin]))
    if spec.version:
        nodes.append(_node("version", args=[spec.version]))
    if spec.about:
        nodes.append(_node("about", args=[spec.about]))
    if spec.long:
        nodes.append(_node("long_about", args=[spec.long]))
    if spec.usage:
        nodes.append(_node("usage", args=[spec.usage]))

    for flag in spec.flags:
        nodes.append(_build_flag_node(flag))
    for arg in spec.args:
        nodes.append(_build_arg_node(arg))
    for cmd in spec.cmds:
        nodes.append(_build_command_node(cmd))

    doc = ckdl.Document(*nodes)
    return str(doc)


def validate_kdl(kdl: str) -> None:
    """Validate KDL output by parsing it back. Throws if invalid."""
    if not kdl.strip():
        return
    ckdl.parse(kdl)
